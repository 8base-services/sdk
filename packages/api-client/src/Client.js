// @flow
import type { DocumentNode } from 'graphql';
import gql from 'graphql-tag';
import { GraphQLClient } from 'graphql-request';
import errorCodes from '@8base/error-codes';
import * as R from 'ramda';

const USER_REFRESH_TOKEN_QUERY = gql`
  mutation UserRefreshToken($refreshToken: String!, $email: String!) {
    userRefreshToken(data: {
      refreshToken: $refreshToken,
      email: $email,
    }) {
      refreshToken
      idToken
    }
  }
`;

const hasTokenExpiredErrorCode = R.any(
  R.propEq('code', errorCodes.TokenExpiredErrorCode),
);

const getRefreshToken = R.path(['userRefreshToken', 'refreshToken']);
const getIdToken = R.path(['userRefreshToken', 'idToken']);

type GraphQLClientResponse = {
  errors: Array<any>,
  data: any,
};

type GraphQLClientRequest = {
  query: DocumentNode,
  variables: Object,
};

type GraphQLClientError = {
  response: GraphQLClientResponse,
  request: GraphQLClientRequest,
};

/**
 * Client provides methods to make requests to the API.
 * @param endpoint - The endpoint which are using for make requests.
 */
class Client {
  gqlc: GraphQLClient;

  accountId: string;
  email: string;
  idToken: ?string;
  refreshToken: string;

  constructor(endpoint: string) {
    this.gqlc = new GraphQLClient(endpoint);
  }

  /**
   * Update id token.
   * @param idToken - The id token.
   */
  setIdToken(idToken: ?string) {
    this.idToken = idToken;

    if (idToken) {
      this.gqlc.setHeader('Authorization', `Bearer ${idToken}`);
    } else {
      delete this.gqlc.options.headers['Authorization'];
    }
  }

  /**
   * Update refresh token.
   * @param refreshToken - The refresh token.
   */
  setRefreshToken(refreshToken: string) {
    this.refreshToken = refreshToken;
  }

  /**
   * Update user email.
   * @param email - The user email.
   */
  setEmail(email: string) {
    this.email = email;
  }

  /**
   * Update account identifier.
   * @param accountId - The account identifier.
   */
  setAccountId(accountId: string) {
    this.accountId = accountId;
    this.gqlc.setHeader('account-id', accountId);
  }

  async tryToRefreshToken(err: GraphQLClientError) {
    const { refreshToken, email } = this;

    this.setIdToken(null);

    const response = await this.request(USER_REFRESH_TOKEN_QUERY, {
      refreshToken,
      email,
    });

    const newRefreshToken = getRefreshToken(response);
    const newIdToken = getIdToken(response);

    this.setRefreshToken(newRefreshToken);
    this.setIdToken(newIdToken);

    return this.request(err.request.query, err.request.variables);
  }

  handleRequestErrors = (err: GraphQLClientError) => {
    if (hasTokenExpiredErrorCode(err.response.errors || [])) {
      return this.tryToRefreshToken(err);
    }

    throw err;
  };

  /**
   * Send request to the API
   * @param query - GraphQL query.
   * @param variables - The variables that will be used when executing the query.
   * @returns {Promise}
   */
  request(query: string | DocumentNode, variables: Object = {}) {
    return this.gqlc.request(query, variables).catch(this.handleRequestErrors);
  }
}

export { Client };
