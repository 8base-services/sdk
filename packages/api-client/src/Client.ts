import * as R from 'ramda';
import { DocumentNode } from 'graphql';
import gql from 'graphql-tag';
import { GraphQLClient } from 'graphql-request';
import errorCodes from '@8base/error-codes';

import { importWorkspace } from './importWorkspace';
import { RefreshTokenInvalidError } from './RefreshTokenInvalidError';

const USER_REFRESH_TOKEN_QUERY = gql`
  mutation UserRefreshToken($refreshToken: String!) {
    userRefreshToken(data: { refreshToken: $refreshToken }) {
      refreshToken
      idToken
    }
  }
`;

const hasIdTokenExpiredError = R.any(
  R.allPass([R.propEq('code', errorCodes.TokenExpiredErrorCode), R.propEq('message', 'Token expired')]),
);

const getRefreshToken: (obj: any) => string = R.path<any>(['userRefreshToken', 'refreshToken']);
const getIdToken: (obj: any) => string = R.path<any>(['userRefreshToken', 'idToken']);

type GraphQLClientResponse = {
  errors: object[];
  data: object;
};

type GraphQLClientRequest = {
  query: DocumentNode;
  variables: object;
};

type GraphQLClientError = {
  response: GraphQLClientResponse;
  request: GraphQLClientRequest;
};

/**
 * Client provides methods to make requests to the API.
 * @param endpoint - The endpoint which are using for make requests.
 */
class Client {
  public gqlc: GraphQLClient;

  public workspaceId?: string;
  public idToken?: string | null;
  public refreshToken?: string;

  constructor(endpoint: string) {
    this.gqlc = new GraphQLClient(endpoint);
  }

  /**
   * Update id token.
   * @param idToken - The id token.
   */
  public setIdToken(idToken: string | null) {
    this.idToken = idToken;

    if (idToken) {
      this.gqlc.setHeader('Authorization', `Bearer ${idToken}`);
    } else {
      // @ts-ignore. Check access to private variable.
      delete this.gqlc.options.headers.Authorization;
    }
  }

  /**
   * Update refresh token.
   * @param refreshToken - The refresh token.
   */
  public setRefreshToken(refreshToken: string) {
    this.refreshToken = refreshToken;
  }

  /**
   * Update workspace identifier.
   * @param workspaceId - The workspace identifier.
   */
  public setWorkspaceId(workspaceId: string) {
    this.workspaceId = workspaceId;
    this.gqlc.setHeader('workspace', workspaceId);
  }

  public async tryToRefreshToken(err: GraphQLClientError) {
    const { refreshToken } = this;

    this.setIdToken(null);

    let response = null;

    try {
      response = await this.gqlc.request(USER_REFRESH_TOKEN_QUERY, {
        refreshToken,
      });
    } catch (err) {
      throw new RefreshTokenInvalidError();
    }

    const newRefreshToken = getRefreshToken(response);
    const newIdToken = getIdToken(response);

    this.setRefreshToken(newRefreshToken);
    this.setIdToken(newIdToken);

    return this.request(err.request.query, err.request.variables);
  }

  public handleRequestErrors = (err: GraphQLClientError) => {
    if (hasIdTokenExpiredError(R.pathOr([], ['response', 'errors'], err))) {
      return this.tryToRefreshToken(err);
    }

    throw err;
  };

  /**
   * Send request to the API
   * @param query - GraphQL query.
   * @param variables - The variables that will be used when executing the query.
   * @returns {Promise}
   */
  public request(query: string | DocumentNode, variables: object = {}): Promise<any> {
    return this.gqlc.request(query, variables).catch(this.handleRequestErrors);
  }

  public importWorkspace(workspace: object) {
    return importWorkspace(this.request.bind(this), workspace);
  }
}

export { Client };
