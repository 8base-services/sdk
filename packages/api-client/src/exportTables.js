// @flow
import type { DocumentNode } from 'graphql';

import { TABLES_LIST_QUERY } from './constants';

type ExportTablesConfig = {
  withSystemTables?: boolean
}

export const exportTables =
  async (request: (query: string | DocumentNode, variables?: Object) => Promise<Object>, config: ExportTablesConfig = {}) => {
    const variables = config.withSystemTables
      ? {}
      : {
        filter: {
          onlyUserTables: true,
        },
      };

    const tablesListData = await request(TABLES_LIST_QUERY, variables);

    return tablesListData.tablesList.items;
  };
