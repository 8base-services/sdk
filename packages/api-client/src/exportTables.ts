// @flow
import { DocumentNode } from 'graphql';
import { TableSchema } from '@8base/utils';

import { TABLES_LIST_QUERY } from './constants';
import { TableSchemaResponse } from './types';

type ExportTablesConfig = {
  withSystemTables?: boolean
}

export const exportTables =
  async (request: <T extends object>(query: string | DocumentNode, variables?: Object) => Promise<T>, config: ExportTablesConfig = {}) => {
    const variables = config.withSystemTables
      ? {}
      : {
        filter: {
          onlyUserTables: true,
        },
      };

    const tablesListData = await request<TableSchemaResponse>(TABLES_LIST_QUERY, variables);

    return tablesListData.tablesList.items;
  };
