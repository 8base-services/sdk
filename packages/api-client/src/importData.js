// @flow
import { formatDataForMutation, isRelationField, getFieldSchemaByName, getTableSchemaByName, isFileField, isListField, MUTATION_TYPE } from '@8base/utils';
import { SchemaNameGenerator } from '@8base/schema-name-generator';
import * as filestack from 'filestack-js';
import * as R from 'ramda';
import type { DocumentNode } from 'graphql';

import { TABLES_LIST_QUERY, USER_QUERY, FILE_UPLOAD_INFO_QUERY } from './constants';

const getRemoteEntityId = (localData: Object, fieldSchema: Object, $id: string, userId: string) => {
  let id = null;

  if ($id === '$currentUserId' && fieldSchema.relation.refTable.name === 'Users') {
    id = userId;
  } else {
    ({ id } = localData[fieldSchema.relation.refTable.name][$id]);
  }

  return id;
};

const MAX_PARALLEL_CREATIONS = 50;

const uploadFiles = async (record, tableSchema, filestackClient, path) => {
  const fieldNames = R.keys(record);

  let nextRecord = record;

  for (let i = fieldNames.length - 1; i >= 0; i--) {
    const fieldName = fieldNames[i];

    const fieldSchema = getFieldSchemaByName(fieldName, tableSchema);

    if (fieldSchema && isFileField(fieldSchema)) {
      if (isListField(fieldSchema)) {
        if (Array.isArray(record[fieldName])) {
          for (let j = 0; j < record[fieldName].length; j++) {
            nextRecord = R.assocPath([fieldName, j, 'fileId'], (await (filestackClient.storeURL(record[fieldName][j].url, {
              path,
            }))).handle, nextRecord);
            nextRecord = R.dissocPath([fieldName, j, 'url'], nextRecord);
          }
        }
      } else {
        if (record[fieldName]) {
          nextRecord = R.assocPath([fieldName, 'fileId'], (await (filestackClient.storeURL(record[fieldName].url, {
            path,
          }))).handle, nextRecord);
          nextRecord = R.dissocPath([fieldName, 'url'], nextRecord);
        }
      }
    }
  }

  return nextRecord;
};

export const importData = async (request: (query: string | DocumentNode, variables?: Object) => Promise<Object>, schemaData: Object) => {
  const { tablesList: { items: tableSchema }} = await request(TABLES_LIST_QUERY, {
    filter: {
      onlyUserTables: false,
    },
  });

  let fileUploadInfo = {};

  try {
    ({ fileUploadInfo } = await request(FILE_UPLOAD_INFO_QUERY));
  } catch (e) {
    // eslint-disable-next-line no-console
    console.log('Can\'t fetch file upload info', e);
  }

  const filestackClient = filestack.init(fileUploadInfo.apiKey, {
    security: {
      policy: fileUploadInfo.policy,
      signature: fileUploadInfo.signature,
    },
  });

  const localData = {};

  for (const tableName of Object.keys(schemaData)) {
    localData[tableName] = {};

    for (let i = 0; i < schemaData[tableName].length / MAX_PARALLEL_CREATIONS; i++) {
      const tempData = schemaData[tableName].slice(i * MAX_PARALLEL_CREATIONS, (i + 1) * MAX_PARALLEL_CREATIONS);

      await Promise.all(tempData.map(async (item) => {
        item = await uploadFiles(item, getTableSchemaByName(tableName, tableSchema), filestackClient, fileUploadInfo.path);

        const data = formatDataForMutation(MUTATION_TYPE.CREATE, tableName, item, tableSchema, {
          skip: (value, fieldSchema) => isRelationField(fieldSchema),
        });

        const fieldData = await request(`
          mutation create($data: ${SchemaNameGenerator.getCreateInputName(tableName)}!) {
            field: ${SchemaNameGenerator.getCreateItemFieldName(tableName)}(data: $data) {
              id
            }
          }
        `, {
          data,
        });

        if (item.$id) {
          localData[tableName][item.$id] = fieldData.field;
        }
      }));
    }
  }

  let userId = '';

  try {
    const userData = await request(USER_QUERY);

    userId = userData.user.id;
  } catch (e) {
    // eslint-disable-next-line no-console
    console.log('Can\'t fetch user info', e);
  }

  for (const tableName of Object.keys(schemaData)) {
    for (const item of schemaData[tableName]) {
      const data = formatDataForMutation(MUTATION_TYPE.UPDATE, tableName, item, tableSchema, {
        skip: (value, fieldSchema) => !isRelationField(fieldSchema),
        mutate: (value, plainValue, fieldSchema) => {
          if (Array.isArray(plainValue)) {
            return {
              connect: plainValue.map(({ $id }) => ({
                id: getRemoteEntityId(localData, fieldSchema, $id, userId),
              })),
            };
          }

          if (!plainValue) {
            return null;
          }

          const id = getRemoteEntityId(localData, fieldSchema, plainValue.$id, userId);

          return { connect: { id }};
        },
      });

      await request(`
        mutation update($data: ${SchemaNameGenerator.getUpdateInputName(tableName)}!) {
          field: ${SchemaNameGenerator.getUpdateItemFieldName(tableName)}(data: $data) {
            id
          }
        }
      `, {
        data: {
          ...data,
          id: localData[tableName][item.$id].id,
        },
      });
    }
  }
};
