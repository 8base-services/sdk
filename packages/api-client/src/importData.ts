import {
  formatDataForMutation,
  tablesListSelectors,
  tableFieldSelectors,
  tableSelectors,
  MUTATION_TYPE,
  FieldSchema,
  TableSchema,
  SDKError,
  ERROR_CODES,
  PACKAGES,
} from '@8base/utils';
import { SchemaNameGenerator } from '@8base/schema-name-generator';
import * as filestack from 'filestack-js';
import * as R from 'ramda';
import { DocumentNode } from 'graphql';

import { TABLES_LIST_QUERY, USER_QUERY, FILE_UPLOAD_INFO_QUERY } from './constants';
import { SchemaResponse } from './types';

const getRemoteEntityId = (
  localData: { [key: string]: any },
  fieldSchema: FieldSchema,
  $id: string,
  userId: string,
) => {
  let id = null;

  if ($id === '$currentUserId' && fieldSchema.relation.refTable.name === 'Users') {
    id = userId;
  } else {
    ({ id } = localData[fieldSchema.relation.refTable.name][$id]);
  }

  return id;
};

const MAX_THREADS = 5;

const uploadFiles = async (record: any, tableSchema: TableSchema, filestackClient: any, path: string) => {
  const fieldNames: string[] = R.keys(record) as any;

  let nextRecord = record;

  for (let i = fieldNames.length - 1; i >= 0; i--) {
    const fieldName = fieldNames[i];

    const fieldSchema = tableSelectors.getFieldByName(tableSchema, fieldName);

    if (fieldSchema && tableFieldSelectors.isFileField(fieldSchema)) {
      if (tableFieldSelectors.isListField(fieldSchema)) {
        if (Array.isArray(record[fieldName])) {
          for (let j = 0; j < record[fieldName].length; j++) {
            nextRecord = R.assocPath(
              [fieldName, j, 'fileId'],
              (await filestackClient.storeURL(record[fieldName][j].url, {
                path,
              })).handle,
              nextRecord,
            );
            nextRecord = R.dissocPath([fieldName, j, 'url'], nextRecord);
          }
        }
      } else {
        if (record[fieldName]) {
          nextRecord = R.assocPath(
            [fieldName, 'fileId'],
            (await filestackClient.storeURL(record[fieldName].url, {
              path,
            })).handle,
            nextRecord,
          );
          nextRecord = R.dissocPath([fieldName, 'url'], nextRecord);
        }
      }
    }
  }

  return nextRecord;
};

type ImportOptions = {
  maxThreads?: number;
};

export const importData = async (
  request: <T extends object>(query: string | DocumentNode, variables?: object) => Promise<T>,
  schemaData: { [key: string]: any },
  options: ImportOptions = {},
) => {
  const {
    tablesList: { items: schema },
  } = await request<SchemaResponse>(TABLES_LIST_QUERY, {
    filter: {
      onlyUserTables: false,
    },
  });

  let fileUploadInfo = {} as any;

  try {
    ({ fileUploadInfo } = await request(FILE_UPLOAD_INFO_QUERY));
  } catch (e) {
    // tslint:disable-next-line no-console
    console.log("Can't fetch file upload info", e);
  }

  const filestackClient = filestack.init(fileUploadInfo.apiKey, {
    security: {
      policy: fileUploadInfo.policy,
      signature: fileUploadInfo.signature,
    },
  });

  const localData = {} as any;

  const maxThreads: number = R.propOr(MAX_THREADS, 'maxThreads', options);

  for (const tableName of Object.keys(schemaData)) {
    localData[tableName] = {};

    const tableSchema = tablesListSelectors.getTableByName(schema, tableName);

    if (!tableSchema) {
      throw new SDKError(ERROR_CODES.TABLE_NOT_FOUND, PACKAGES.API_CLIENT, `Table with name ${tableName} not found`);
    }

    for (let i = 0; i < schemaData[tableName].length / maxThreads; i++) {
      const tempData = schemaData[tableName].slice(i * maxThreads, (i + 1) * maxThreads);

      await Promise.all(
        tempData.map(async (item: any) => {
          item = await uploadFiles(item, tableSchema, filestackClient, fileUploadInfo.path);

          const data = formatDataForMutation({
            type: MUTATION_TYPE.CREATE,
            tableName,
            data: item,
            schema,
            options: {
              skip: (value: any, fieldSchema: FieldSchema) => tableFieldSelectors.isRelationField(fieldSchema),
            }
          });

          const fieldData = await request<{ field: { id: string } }>(
            `
          mutation create($data: ${SchemaNameGenerator.getCreateInputName(tableName)}!) {
            field: ${SchemaNameGenerator.getCreateItemFieldName(tableName)}(data: $data) {
              id
            }
          }
        `,
            {
              data,
            },
          );

          if (item.$id) {
            localData[tableName][item.$id] = fieldData.field;
          }
        }),
      );
    }
  }

  let userId = '';

  try {
    const userData = await request<{ user: { id: string } }>(USER_QUERY);

    userId = userData.user.id;
  } catch (e) {
    // tslint:disable-next-line no-console
    console.log(`Can't fetch user info`, e);
  }

  for (const tableName of Object.keys(schemaData)) {
    for (const item of schemaData[tableName]) {
      const data = formatDataForMutation({
        type: MUTATION_TYPE.UPDATE,
        tableName,
        data: item,
        schema,
        options: {
          mutate: (value: any, plainValue: any, fieldSchema: FieldSchema) => {
            if (Array.isArray(plainValue)) {
              return {
                connect: plainValue.map(({ $id }) => ({
                  id: getRemoteEntityId(localData, fieldSchema, $id, userId),
                })),
              };
            }

            if (!plainValue) {
              return null;
            }

            const id = getRemoteEntityId(localData, fieldSchema, plainValue.$id, userId);

            return { connect: { id } };
          },
          skip: (value: any, fieldSchema: FieldSchema) => !tableFieldSelectors.isRelationField(fieldSchema),
        }
      });

      await request(
        `
        mutation update($data: ${SchemaNameGenerator.getUpdateInputName(tableName)}!) {
          field: ${SchemaNameGenerator.getUpdateItemFieldName(tableName)}(data: $data) {
            id
          }
        }
      `,
        {
          data: {
            ...data,
            id: localData[tableName][item.$id].id,
          },
        },
      );
    }
  }
};
