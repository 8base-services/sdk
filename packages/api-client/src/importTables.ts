// @flow
import * as R from 'ramda';
import errorCodes from '@8base/error-codes';
import type { Schema } from '@8base/utils';
import type { DocumentNode } from 'graphql';

import { TABLES_LIST_QUERY, TABLE_CREATE_MUTATION, FIELD_CREATE_MUTATION } from './constants';

type ImportTablesOptions = {
  debug?: boolean,
};

const handleError = (message: string, e: Object, debug: boolean) => {
  // eslint-disable-next-line no-console
  console.warn(message);

  if (debug) {
    // eslint-disable-next-line no-console
    console.warn(JSON.stringify(e.response, null, 2));
  } else if (R.pathEq(['response', 'errors', 0, 'code'], errorCodes.ValidationErrorCode, e)) {
    // eslint-disable-next-line no-console
    console.warn(JSON.stringify(R.path(['response', 'errors', 0, 'details'], e), null, 2));
  } else {
    // eslint-disable-next-line no-console
    console.warn(R.path(['response', 'errors', 0, 'message'], e));
  }
};

export const importTables = async (
  request: (query: string | DocumentNode, variables?: Object) => Promise<Object>,
  schema: Schema,
  options: ImportTablesOptions = {},
) => {
  const debug = R.propOr(false, 'debug', options);

  for (const table of schema) {
    try {
      await request(TABLE_CREATE_MUTATION, {
        data: {
          name: table.name,
          displayName: table.displayName,
        },
      });
    } catch (e) {
      handleError(`\nCan't create table "${table.name}"`, e, debug);
    }
  }

  const tablesListData = await request(TABLES_LIST_QUERY, {
    filter: {
      onlyUserTables: false,
    },
  });

  const tablesByName = R.indexBy(R.prop('name'), tablesListData.tablesList.items);

  for (const schemaTable of schema) {
    if (schemaTable.isSystem) {
      continue;
    }

    const table = tablesByName[schemaTable.name];

    for (const schemaField of schemaTable.fields) {
      if (schemaField.isSystem) {
        continue;
      }

      const field = R.pick([
        'name',
        'displayName',
        'fieldType',
        'isList',
        'isRequired',
        'isUnique',
        'defaultValue',
        'description',
        'fieldTypeAttributes',
        'relation',
        'tableId',
      ], schemaField);

      if (field.fieldTypeAttributes) {
        field.fieldTypeAttributes = R.pick([
          'format',
          'fieldSize',
          'listOptions',
          'precision',
          'currency',
          'minValue',
          'maxValue',
          'maxSize',
          'typeRestrictions',
          'isBigInt',
        ], field.fieldTypeAttributes);
      }

      if (field.relation) {
        const refTableName = field.relation.refTable.name;

        const refTable = tablesByName[refTableName];

        if (refTable.relations && refTable.relations[schemaTable.name]) {
          if (refTable.relations[schemaTable.name].includes(field.name)) {
            continue;
          }
        }

        field.relation = {
          ...R.pick([
            'refFieldIsList',
            'refFieldIsRequired',
            'refFieldName',
            'refFieldDisplayName',
          ], field.relation),
          refTableId: refTable.id,
        };

        table.relations = table.relations || {};

        if (table.relations[refTableName]) {
          table.relations[refTableName].push(field.relation.refFieldName);
        } else {
          table.relations[refTableName] = [field.relation.refFieldName];
        }
      }

      field.tableId = table.id;

      try {
        await request(FIELD_CREATE_MUTATION, {
          data: field,
        });
      } catch (e) {
        handleError(`\nCan't create field "${field.name}"`, e, debug);
      }
    }
  }
};
