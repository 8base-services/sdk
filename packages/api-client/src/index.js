export { importTables } from './importTables';
export { importData } from './importData';
export { exportTables } from './exportTables';
export { Client } from './Client';
