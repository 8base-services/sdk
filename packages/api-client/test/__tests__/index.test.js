import gql from 'graphql-tag';
import errorCodes from '@8base/error-codes';

import { Client } from '../../src';

it('As a developer, I can create client and send request.', async () => {
  const requestPromise = mockRequest('https://api.test.8base.com');

  const client = new Client('https://api.test.8base.com');

  await client.request('query { companyName }');

  const { body } = await requestPromise;

  expect(body).toMatchSnapshot();
});

it('As a developer, I can create client and send request with variables.', async () => {
  const requestPromise = mockRequest('https://api.test.8base.com');

  const client = new Client('https://api.test.8base.com');

  await client.request('query { companyName }', { variable: 2 });

  const { body } = await requestPromise;

  expect(body).toMatchSnapshot();
});

it('As a developer, I can create client, set API credentials and send request.', async () => {
  const requestPromise = mockRequest('https://api.test.8base.com');

  const client = new Client('https://api.test.8base.com');

  client.setIdToken('idToken');
  client.setAccountId('accountId');

  await client.request('query { companyName }');

  const request = await requestPromise;

  expect(request).toMatchSnapshot();
});

it('As a developer, I can send queries with graphql tag.', async () => {
  const requestPromise = mockRequest('https://api.test.8base.com');

  const client = new Client('https://api.test.8base.com');

  await client.request(gql`query { companyName }`);

  const request = await requestPromise;

  expect(request).toMatchSnapshot();
});

it('When client receive token expired error, it should refresh token and repeat my request.', async () => {
  mockRequest('https://api.test.8base.com', 502, {
    errors: [{
      code: errorCodes.TokenExpiredErrorCode,
    }],
    data: null,
  });

  mockRequest('https://api.test.8base.com', 200, {
    data: {
      system: {
        userRefreshToken: {
          refreshToken: 'newRefreshToken',
          idToken: 'newIdToken',
        },
      },
    },
  });

  const requestPromise = mockRequest('https://api.test.8base.com');

  const client = new Client('https://api.test.8base.com');

  client.setIdToken('idToken');
  client.setAccountId('accountId');
  client.setEmail('test@site.com');

  await client.request('query { companyName }');

  const request = await requestPromise;

  expect(request).toMatchSnapshot();
});

it('When client receive other errors, it should throw that error.', async () => {
  mockRequest('https://api.test.8base.com', 502, {
    errors: [{
      code: errorCodes.InvalidTokenErrorCode,
    }],
    data: null,
  });

  mockRequest('https://api.test.8base.com');

  const client = new Client('https://api.test.8base.com');

  client.setIdToken('idToken');
  client.setAccountId('accountId');
  client.setEmail('test@site.com');

  let error = null;

  try {
    await client.request(gql`query { companyName }`);
  } catch (err) {
    error = err;
  }

  expect(error.response).toMatchSnapshot();
});
