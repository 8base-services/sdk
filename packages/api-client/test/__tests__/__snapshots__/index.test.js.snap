// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`As a developer, I can create client and send request with variables. 1`] = `
Object {
  "query": "query { companyName }",
  "variables": Object {
    "variable": 2,
  },
}
`;

exports[`As a developer, I can create client and send request. 1`] = `
Object {
  "query": "query { companyName }",
  "variables": Object {},
}
`;

exports[`As a developer, I can create client, set API credentials and send request. 1`] = `
Object {
  "body": Object {
    "query": "query { companyName }",
    "variables": Object {},
  },
  "headers": Object {
    "accept": Array [
      "*/*",
    ],
    "accept-encoding": Array [
      "gzip,deflate",
    ],
    "authorization": Array [
      "Bearer idToken",
    ],
    "connection": Array [
      "close",
    ],
    "content-length": Array [
      "48",
    ],
    "content-type": Array [
      "application/json",
    ],
    "host": "api.test.8base.com",
    "user-agent": Array [
      "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
    ],
    "workspace": Array [
      "workspaceId",
    ],
  },
}
`;

exports[`As a developer, I can send queries with graphql tag. 1`] = `
Object {
  "body": Object {
    "query": Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "arguments": Array [],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "companyName",
                },
              },
            ],
          },
          "variableDefinitions": Array [],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 21,
        "start": 0,
      },
    },
    "variables": Object {},
  },
  "headers": Object {
    "accept": Array [
      "*/*",
    ],
    "accept-encoding": Array [
      "gzip,deflate",
    ],
    "connection": Array [
      "close",
    ],
    "content-length": Array [
      "326",
    ],
    "content-type": Array [
      "application/json",
    ],
    "host": "api.test.8base.com",
    "user-agent": Array [
      "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
    ],
  },
}
`;

exports[`When client receive other errors, it should throw that error. 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "code": "InvalidTokenError",
    },
  ],
  "status": 502,
}
`;

exports[`When client receive token expired error, it should refresh token and repeat my request. 1`] = `
Object {
  "body": Object {
    "query": Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "UserRefreshToken",
          },
          "operation": "mutation",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "data",
                    },
                    "value": Object {
                      "fields": Array [
                        Object {
                          "kind": "ObjectField",
                          "name": Object {
                            "kind": "Name",
                            "value": "refreshToken",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "refreshToken",
                            },
                          },
                        },
                        Object {
                          "kind": "ObjectField",
                          "name": Object {
                            "kind": "Name",
                            "value": "email",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "email",
                            },
                          },
                        },
                      ],
                      "kind": "ObjectValue",
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "userRefreshToken",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "refreshToken",
                      },
                    },
                    Object {
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "idToken",
                      },
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "refreshToken",
                },
              },
            },
            Object {
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "email",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 209,
        "start": 0,
      },
    },
    "variables": Object {
      "email": "test@site.com",
    },
  },
  "headers": Object {
    "accept": Array [
      "*/*",
    ],
    "accept-encoding": Array [
      "gzip,deflate",
    ],
    "connection": Array [
      "close",
    ],
    "content-length": Array [
      "1418",
    ],
    "content-type": Array [
      "application/json",
    ],
    "host": "api.test.8base.com",
    "user-agent": Array [
      "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
    ],
    "workspace": Array [
      "workspaceId",
    ],
  },
}
`;

exports[`When client receive token expired error, it should refresh token and repeat my request. 2`] = `
Object {
  "body": Object {
    "query": "query { companyName }",
    "variables": Object {},
  },
  "headers": Object {
    "accept": Array [
      "*/*",
    ],
    "accept-encoding": Array [
      "gzip,deflate",
    ],
    "authorization": Array [
      "Bearer newIdToken",
    ],
    "connection": Array [
      "close",
    ],
    "content-length": Array [
      "48",
    ],
    "content-type": Array [
      "application/json",
    ],
    "host": "api.test.8base.com",
    "user-agent": Array [
      "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
    ],
    "workspace": Array [
      "workspaceId",
    ],
  },
}
`;
